File: Engine/ceCore/entity/transform.cc

#include <ceCore/entity/transform.hh>
#include <ceCore/entity/spatialstate.hh>

namespace cryo
{


Transform::Transform(SpatialState *state, const Matrix4f &matrix, const Matrix4f &parentMatrix)
    : m_state(state), m_matrix(matrix), m_translation(0.0f, 0.0f, 0.0f), m_rotation(0.0f, 0.0f, 0.0f, 1.0f), m_scale(
    1.0f,
    1.0f,
    1.0f), m_dirty(true), m_parentMatrix(parentMatrix)
{
  ExtractTRS();
}

void Transform::ExtractTRS()
{
  m_translation = m_matrix.GetTranslation();
  m_scale       = Vector3f(m_matrix.GetXAxis().Length(),
                           m_matrix.GetYAxis().Length(),
                           m_matrix.GetZAxis().Length());

  Matrix3f rotMat(m_matrix.m00 / m_scale.x, m_matrix.m01 / m_scale.x, m_matrix.m02 / m_scale.x,
                  m_matrix.m10 / m_scale.y, m_matrix.m11 / m_scale.y, m_matrix.m12 / m_scale.y,
                  m_matrix.m20 / m_scale.z, m_matrix.m21 / m_scale.z, m_matrix.m22 / m_scale.z
  );
  m_rotation = Quaternion::FromMatrix(rotMat);
}

Transform &Transform::SetTranslation(const Vector3f &translation)
{
  m_translation = translation;
  m_dirty       = true;
  return *this;
}

Transform &Transform::SetTranslation(float x, float y, float z)
{
  m_translation.x = x;
  m_translation.y = y;
  m_translation.z = z;
  m_dirty = true;
  return *this;
}

const Vector3f &Transform::GetTranslation() const
{
  return m_translation;
}


Transform &Transform::SetRotation(const Quaternion &rotation)
{
  m_rotation = rotation;
  m_dirty    = true;
  return *this;
}

const Quaternion &Transform::GetRotation() const
{
  return m_rotation;
}


Transform &Transform::SetScale(const Vector3f &scale)
{
  m_scale = scale;
  m_dirty = true;
  return *this;
}


const Vector3f &Transform::GetScalar() const
{
  return m_scale;
}

Transform &Transform::LookAt(const Vector3f &at, const Vector3f &up)
{
  Vector3f from = Matrix4f::Transform(m_parentMatrix, m_translation);

  Matrix4f matrix;
  matrix.SetLookAtInv(from, at, up);

  Matrix4f parent = m_parentMatrix;
  parent.Invert();
  m_matrix = matrix * parent;

  ExtractTRS();
  m_dirty = true;
  return *this;
}


const Matrix4f &Transform::GetMatrix() const
{
  UpdateMatrix();
  return m_matrix;
}


const Matrix4f &Transform::GetGlobalMatrix() const
{
  UpdateMatrix();
  return m_globalMatrix;
}

Vector3f Transform::GetForward() const
{
  UpdateMatrix();
  return m_globalMatrix * Vector3f(0, 0, 1);
}


Vector3f Transform::GetBackward() const
{
  UpdateMatrix();
  return m_globalMatrix * Vector3f(0, 0, -1);
}


Vector3f Transform::GetLeft() const
{
  UpdateMatrix();
  return m_globalMatrix * Vector3f(-1, 0, 0);
}


Vector3f Transform::GetRight() const
{
  UpdateMatrix();
  return m_globalMatrix * Vector3f(1, 0, 0);
}


Vector3f Transform::GetUp() const
{
  UpdateMatrix();
  return m_globalMatrix * Vector3f(0, 1, 0);
}


Vector3f Transform::GetDown() const
{
  UpdateMatrix();
  return m_globalMatrix * Vector3f(0, -1, 0);
}

void Transform::Finish()
{
  UpdateMatrix();
  m_state->SetLocalMatrix(m_matrix);
}


void Transform::UpdateMatrix() const
{
  if (m_dirty)
  {
    m_matrix = m_rotation.ToMatrix4();
    m_matrix.SetTranslation(m_translation);
    m_matrix.Scale(m_scale);
    m_globalMatrix = m_parentMatrix * m_matrix;
    m_dirty        = false;
  }

}

}

File: JCryoStorm/CsCore/src/main/java/org/cryo/core/entity/Transform.java
package org.cryo.core.entity;

import org.cryo.core.math.Matrix3f;
import org.cryo.core.math.Matrix4f;
import org.cryo.core.math.Quaternion;
import org.cryo.core.math.Vector3f;

public class Transform {

    private final SpatialState spatialState;
    private final Vector3f translation = new Vector3f();
    private final Quaternion rotation = new Quaternion();
    private final Vector3f scale = new Vector3f(1, 1, 1);

    private final Matrix4f parentMatrix = new Matrix4f();
    private boolean dirty;
    private final Matrix4f matrix = new Matrix4f();
    private final Matrix4f globalMatrix = new Matrix4f();


    public Transform(SpatialState spatialState) {
        this.spatialState = spatialState;
    }


    public Transform set(Matrix4f matrix,
                         Matrix4f parentMatrix) {
        this.matrix.set(matrix);
        this.parentMatrix.set(parentMatrix);

        extractTRS();
        return this;
    }

    private void extractTRS() {
        matrix.getTranslation(this.translation);
        scale.set(matrix.getXAxis(new Vector3f()).length(),
                  matrix.getYAxis(new Vector3f()).length(),
                  matrix.getZAxis(new Vector3f()).length());

        Matrix3f rotMat = new Matrix3f(matrix.m00 / scale.x, matrix.m01 / scale.x, matrix.m02 / scale.x,
                                       matrix.m10 / scale.y, matrix.m11 / scale.y, matrix.m12 / scale.y,
                                       matrix.m20 / scale.z, matrix.m21 / scale.z, matrix.m22 / scale.z);
        rotation.set(Quaternion.fromMatrix(rotMat));
    }

    public Transform setTranslation(Vector3f translation) {
        this.translation.set(translation);
        dirty = true;
        return this;
    }

    public Transform setTranslation(float x, float y, float z) {
        translation.set(x, y, z);
        dirty = true;
        return this;
    }

    public Vector3f getTranslation(Vector3f out) {
        out.set(translation);
        return out;
    }


    public Transform setRotation(Quaternion rotation) {
        this.rotation.set(rotation);
        dirty = true;
        return this;
    }


    public Quaternion getTranslation(Quaternion out) {
        out.set(rotation);
        return out;
    }

    public Transform setScale(Vector3f scale) {
        this.scale.set(scale);
        dirty = true;
        return this;
    }

    public Transform setScale(float x, float y, float z) {
        this.scale.set(x, y, z);
        dirty = true;
        return this;
    }

    public Vector3f getScale(Vector3f out) {
        out.set(scale);
        return out;
    }

    public Transform lookAt(Vector3f at, Vector3f up) {

        Vector3f from = this.parentMatrix.transform(this.translation, new Vector3f());


        Matrix4f tmpMatrix = new Matrix4f();
        tmpMatrix.setLookAtInv(from, at, up);

        Matrix4f parent = new Matrix4f(parentMatrix);
        parent.invert();

        tmpMatrix.mul(parent, this.matrix);

        extractTRS();
        dirty = true;
        return this;
    }


    public Matrix4f getMatrix() {
        updateMatrix();
        return matrix;
    }


    public Matrix4f getGlobalMatrix() {
        updateMatrix();
        return globalMatrix;
    }

    public Vector3f getForward() {
        updateMatrix();
        return globalMatrix.mul(Vector3f.Z, new Vector3f());
    }

    public Vector3f getBackward() {
        updateMatrix();
        return globalMatrix.mul(Vector3f.NEG_Z, new Vector3f());
    }

    public Vector3f getLeft() {
        updateMatrix();
        return globalMatrix.mul(Vector3f.NEG_X, new Vector3f());
    }

    public Vector3f getRight() {
        updateMatrix();
        return globalMatrix.mul(Vector3f.X, new Vector3f());
    }


    public Vector3f getUp() {
        updateMatrix();
        return globalMatrix.mul(Vector3f.Y, new Vector3f());
    }

    public Vector3f getDown() {
        updateMatrix();
        return globalMatrix.mul(Vector3f.NEG_Y, new Vector3f());
    }


    public void finish() {
        updateMatrix();
        this.spatialState.setLocalMatrix(this.matrix);
    }


    void updateMatrix() {
        if (dirty) {
            this.rotation.toMatrix4(this.matrix);
            this.matrix.setTranslation(this.translation);
            this.matrix.scale(this.scale);

            this.globalMatrix.set(parentMatrix).mul(this.matrix, globalMatrix);
            this.dirty = false;
        }
    }


}
